### PR TITLE
Move some common definitions to streams/common.h

### PR DESCRIPTION
**Pull request by jasnell**
Long awaited cleanups and refinements to streams

@harrishancock ... it'll definitely be best and easiest to review this commit-by-commit. You should start to pick up on the common theme here of simplifying.

### CODE REVIEW

File: src/workerd/api/streams/standard.c++ [hunk 1]
```diff
@@ -983,15 +970,11 @@ void ReadableImpl<Self>::pullIfNeeded(jsg::Lock& js, jsg::Ref<Self> self) {
...
+    pulling = false;
     doError(js, kj::mv(reason));
   });
 
-  algorithms.pulling = maybeRunAlgorithm(js,
-                                         algorithms.pull,
-                                         kj::mv(onSuccess),
-                                         kj::mv(onFailure),
-                                         self.addRef());
+  maybeRunAlgorithm(js, algorithms.pull, kj::mv(onSuccess), kj::mv(onFailure), self.addRef());
```
Comment (mikea) [inline src/workerd/api/streams/standard.c++:973]: what was the reason to store the resulting promise before? Won't this abandon and cancel it now?

File: src/workerd/api/streams/standard.h [hunk 2]
```diff
@@ -119,45 +119,31 @@ namespace workerd::api {
...
-//
 // All write operations on a JavaScript-backed WritableStream are processed within the
 // isolate lock using JavaScript promises instead of kj::Promises.
 
 class ReadableStreamJsController;
 class WritableStreamJsController;
 
+template <typename T>
+class WeakRef: public kj::Refcounted {
+  // Used to allow the TransformStreamDefaultController to hold safe
```
Comment (mikea) [inline src/workerd/api/streams/standard.h]: weak in what sense, v8? It is not clear to me what this actually does.

File: src/workerd/api/streams/standard.h [hunk 3]
```diff
@@ -375,7 +355,7 @@ class WritableImpl {
 
   struct Writable {};
 
-  kj::Maybe<WriterOwner&> owner;
+  kj::Own<WeakRef<WritableStreamJsController>> owner;
```
Comment (mikea) [inline src/workerd/api/streams/standard.h:366]: this type is mentioned a lot. maybe typedef would be convenient?

File: src/workerd/api/streams/readable.c++ [hunk 4]
```diff
@@ -380,15 +380,12 @@ jsg::Ref<ReadableStream> ReadableStream::pipeThrough(
 
   auto options = kj::mv(maybeOptions).orDefault({});
   options.pipeThrough = true;
-  maybePipeThrough = controller.pipeTo(js, destination, kj::mv(options)).then(js,
-      JSG_VISITABLE_LAMBDA((this, self = JSG_THIS), (self), (jsg::Lock& js) {
-    maybePipeThrough = nullptr;
+  controller.pipeTo(js, destination, kj::mv(options)).then(js,
```
Comment (mikea) [inline src/workerd/api/streams/readable.c++:383]: again, not sure why it is ok not to hold the ref

**Review by mikea — COMMENTED** [review]
I looked through first 10 changes (up to "No need to hold a reference to the pipe to promise")

Looks good so far. I've got several question regarding promise references which I don't really get.

File: src/workerd/api/streams/standard.c++ (hunk 1 above)
Comment (jasnell) [reply src/workerd/api/streams/standard.c++:973]: JS promises do not get canceled the way kj promises do. Holding these before was defensive only but it's not necessary.

**Review by ohodson — APPROVED** [review]

**Review by harrishancock — APPROVED** [review]
Nice cleanup!

**Comment by jasnell** [comment]
Rebased. The diff got out of sync in a few important ways :-/

File: src/workerd/api/streams/standard.h (hunk 3 above)
Comment (jasnell) [reply src/workerd/api/streams/standard.h:366]: Perhaps, I'll save that for a future cleanup tho

**Pull request merged**